src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -64,10 +69,68 @@ void parallel_reduce (const TeamMember& team,
 
    result = local_tmp;
   } else {
+    const ValueType initial(result);
     Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team, begin, end), lambda, result);
+    result += initial;
   }
 }
 
+/*
+ * Computes a reduction over the view described by 'input' for entries 'input([begin,end))'
+ * The variable 'input' should be an operator s.t. 'input(k)' returns a Pack of size 'vector_size'.
+ * The variable 'total_size' represents the total number of 'ValueType' entries in the view. The
+ * computed reduction is added to 'result'.
+ */
+template <bool Serialize, typename TeamMember, typename InputProvider, typename ValueType>
+static KOKKOS_INLINE_FUNCTION
+void view_reduction (const TeamMember& team,
+                     const int total_size,
```

Comment:
I think we can omit `total_size` because `typename decltype(InputProvider)::n`, `begin`, and `end` fully specify how to deal with the packs.

Reply:
Uhm, InputProvider is not a Pack though. In most cases it'll be a view of packs, but it is also possible (and in Homme it happens) that it is a lambda (this allows to reduce the result of a formula without storing the temporary in a view).

Reply:
Sorry, I should have written `typename decltype(input(0))::n`. That's probably still not quite right, as I'm just typing this in github and not actually test-compiling it. But that will work regardless of `input`'s type, as long as it returns a pack as required.

Reply:
@bartgol my hasty C++-goof aside, don't you agree that neither total_size nor vector_size are needed?

Reply:
For sure vector size. On total_size: what if begin is not at the start of a pack? If begin is always at an aligned position, then yes, I think total_size is not needed. But I have to think about the general case.

Reply:
Although, if begin is not necessarily at the start of a pack, then we need special handling also for the 1st pack...

Reply:
A mid ground solution is to assume that the user "pads" his/her views, so that the use cases of view_reduction will always have begin at an aligned position (kind of like Hommexx did for ppm remap).

Reply:
That's the whole point of `begin` and `end`: to do special handling here instead of the app. We can assume the app will have aligned columns, i.e., `a(0)[0]` is aligned. Then we can conclude that `scalaraize(a)[[begin,end)]` is valid, from which we can work out the padding here. No `total_size` needed.

Reply:
Re: "some sort of check": I suggest we do that in some other PR. A bunch of stuff, like `scalarize` and `smallize`, make quite appropriate assumptions about alignment. Don't forget that a `Pack` _always_ aligns properly, so anything composed of packs will align too. That makes most or all of these assumptions safe. In principle someone could make a `Pack<7>` on a 512-bit-vector-length machine, but I think we shouldn't worry too much about that.

Reply:
You're right, we can work out the padding: if `begin%Pack::n!=0`, then we need padding. Simple check. If we want to be fancy, we could even check in debug mode that `scalarize(a).data()-begin` is at an aligned position (to make sure the user didn't mess up).

@tcclevenger This will require a bit more code to handle the 1st pack, but it should pay off in simplicity on the customer side.

Reply:
Ahh I see now. As I understood the implementation [here](https://github.com/E3SM-Project/E3SM/blob/b7b269b3cfaca869e861fa01e6d65e201dc588a3/components/homme/src/share/cxx/ColumnOps.hpp#L474), they run `parallel_reduce` over `LOOP_LENGTH=NUM_PACKS`, _but_ they do not allow for a custom range, and  instead always reduce the entire view.  If we do allow for a custom range, `begin` and `end` are scalar indices, not pack indices as I have it.  If that is the case, then I agree `begin`, `end`, and `typename decltype(input(0))::n` would be sufficient (assuming `input(0)[0]` is always aligned), but we'll see if I run into any trouble implementing. If `begin` and `end` are pack indices, then I'm not so sure.

Reply:
Yeah, in Hommexx I was free to impl only what hommexx needs. Ekat, however, needs to be as general as possible (within reasonable limits) to accommodate potential needs of E3SM apps.

I think Andrew is right, and we can always assume the view to be aligned. The only scenario where it would not be aligned is if someone allocates a view of plain Reals, then subviews it at a non-aligned index, and reinterpret it as a packed view. I believe we can assume that won't happen (just like C++ in general: the language will always let you free to shoot yourself in the foot).

Reply:
Is the reduction always going to be over contiguous indices? Or do you foresee a situation where the user needs a reduction over, for example, the first entry in each pack? Or some arbitrary set of indices?

Reply:
I think it's very safe to assume reductions will always be over a contiguous set of entries.

Reply:
Let's keep it as [begin,end), contiguous.

Reply:
Just to confirm: begin, end should be scalar indices.